### PR TITLE
Implement support for libvirt provider

### PR DESCRIPTION
**Pull request by zvfvrv**
this PR add the support of libvirt as a Vagrant provider

### CODE REVIEW

**Comment by zvfvrv** [comment]
the ansible log that came out during the build of the base box starting from  "centos/7"

`base-box-builder.k8s-play.local: TASK [Gathering Facts] *********************************************************
    base-box-builder.k8s-play.local: fatal: [base-box-builder.k8s-play.local]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Warning: Permanently added 'base-box-builder.k8s-play.local' (ECDSA) to the list of known hosts.\r\nvagrant@base-box-builder.k8s-play.local: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).", "unreachable": true}
    base-box-builder.k8s-play.local:    to retry, use: --limit @/etc/ansible/playbooks/kubernetes.retry
    base-box-builder.k8s-play.local: 
    base-box-builder.k8s-play.local: PLAY RECAP *********************************************************************
    base-box-builder.k8s-play.local: base-box-builder.k8s-play.local : ok=0    changed=0    unreachable=1    failed=0 `

then the problem exist also during the base box build process

**Comment by ferrarimarco** [comment]
Ok, so the SSH connection is working I guess.

Can you check which public keys the `vagrant` user on the VM has?

**Comment by zvfvrv** [comment]
yes i am able to login trough ssh with the command
`$ vagrant ssh base-box-builder.k8s-play.local`

indeed the key used is the `insecure_private_key` of vagrant 
and the command `$ vagrant ssh-config` output is the following:

Host base-box-builder.k8s-play.local
  HostName 192.168.121.77
  User vagrant
  Port 22
  UserKnownHostsFile /dev/null
  StrictHostKeyChecking no
  PasswordAuthentication no
  IdentityFile /home/francesco/.vagrant.d/insecure_private_key
  IdentitiesOnly yes
  LogLevel FATA

**Comment by ferrarimarco** [comment]
Can open an SSH connection to the box and inspect sshd configuration and allowed keys from there?

**Comment by zvfvrv** [comment]
the authorized keys are the following

`[vagrant@base-box-builder ~]$ cat .ssh/authorized_keys 
ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA6NF8iallvQVp22WDkTkyrtvp9eWW6A8YVr+kz4TjGYe7gHzIw+niNltGEFHzD8+v1I2YJ6oXevct1YeS0o9HZyN1Q9qgCgzUFtdOKLv6IedplqoPkcmF0aYet2PkEDo3MlTBckFXPITAMzF8dJSIFo9D8HfdOV0IAdx4O7PtixWKn5y2hMNG0zQPyUecp4pzC6kivAIhyfHilFR61RGL+GPXQ2MWZWFYbAGjyiYJnAmCP3NOTd0jMZEnDkbUvxhMmBYSdETk1rRgm+R4LOzFUGaHqHDLKLX+FIPKcF96hrucXzcWyLbIbEgE98OHlnVYCzRdK8jlqm8tehUc9c9WhQ== vagrant insecure public key`

and the sshd_config attahced here:
[base-box-builder_sshd_config.txt](https://github.com/ferrarimarco/kubernetes-playground/files/4325353/base-box-builder_sshd_config.txt)

where i can see that `AuthorizedKeysFile      .ssh/authorized_keys` and `PasswordAuthentication no`

**Comment by ferrarimarco** [comment]
Ok, I've digged a bit.

We're currently using password authentication for Ansible (see `default_group_vars` in `Vagrantfile`). We currently use `user: vagrant, password: vagrant` to authenticate, so for now it's a requirement. So, no need to create a customized box.

Reason:
I probably thought it was not critical to have a key, because it's a development playground. I guess it can stay the same. Additionally, we should make sure password authentication is enabled in all cases, not just with the box you're using for libvirt. I'll open another issue to track this.

File: Vagrantfile [hunk 1]
```diff
@@ -43,8 +43,10 @@ BROADCAST_ADDRESS = n | (~n.instance_variable_get(:@mask_addr) & IPAddr::IN4MASK
 CLUSTER_IP_CIDR = settings["pod_network"]["cluster_ip_cidr"]
 SERVICE_IP_CIDR = settings["pod_network"]["service_ip_cidr"]
 
+VAGRANT_PROVIDER = settings["conf"]["vagrant_provider"]
+
 # Vagrant boxes
-VAGRANT_X64_KUBERNETES_NODES_BASE_BOX_ID = settings["conf"]["kubernetes_nodes_base_box_id"]
+VAGRANT_X64_KUBERNETES_NODES_BASE_BOX_ID = settings["conf"]["kubernetes_nodes_base_box_id_"+VAGRANT_PROVIDER]
```
Comment (ferrarimarco) [inline Vagrantfile]: Use an `if` here, it's clearer.

File: Vagrantfile [hunk 2]
```diff
@@ -43,8 +43,10 @@ BROADCAST_ADDRESS = n | (~n.instance_variable_get(:@mask_addr) & IPAddr::IN4MASK
 CLUSTER_IP_CIDR = settings["pod_network"]["cluster_ip_cidr"]
 SERVICE_IP_CIDR = settings["pod_network"]["service_ip_cidr"]
 
+VAGRANT_PROVIDER = settings["conf"]["vagrant_provider"]
```
Comment (ferrarimarco) [inline Vagrantfile]: You need to add the instructions in the README about installing vagrant-libvirt and its dependencies

File: Vagrantfile [hunk 3]
```diff
@@ -319,24 +321,32 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
+          vb.customize ["modifyvm", :id, "--hwvirtex", "on"]
+          vb.customize ["modifyvm", :id, "--memory", info[:mem]]
+          vb.customize ["modifyvm", :id, "--name", hostname]
+          vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
+          vb.customize ["modifyvm", :id, "--natdnshostresolver2", "on"]
+          vb.gui = info[:show_gui]
+          vb.name = hostname
+        end
+      elsif(VAGRANT_PROVIDER == 'libvirt')
+        host.vm.provider :libvirt do |libvirt|
```
Comment (ferrarimarco) [inline Vagrantfile]: I think you should enable nested virtualization for libvirt as well.

File: Vagrantfile [hunk 4]
```diff
@@ -319,24 +321,32 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
       end
-
       host.vm.hostname = hostname
       if(hostname.include? BASE_BOX_BUILDER_VM_NAME)
         host.vm.provision "shell" do |s|
           s.path = "scripts/linux/install-docker.sh"
           s.args = ["--user", "vagrant"]
         end
-
+        if(VAGRANT_PROVIDER == 'libvirt')
```
Comment (ferrarimarco) [inline Vagrantfile]: I think we should do this for the other providers as well. Tracking this in #74

File: defaults.yaml [hunk 5]
```diff
@@ -10,7 +10,9 @@ conf:
   minion_1_name: k8s-minion-1
   minion_2_name: k8s-minion-2
   minion_3_name: k8s-minion-3
-  kubernetes_nodes_base_box_id: "bento/centos-7.4"
+  vagrant_provider: libvirt
+  kubernetes_nodes_base_box_id_virtualbox: "bento/centos-7.4"
```
Comment (ferrarimarco) [inline defaults.yaml]: Nest the values, it's clearer. Example: `kubernetes_nodes_base_box_id.virtualbox`

File: defaults.yaml [hunk 6]
```diff
@@ -10,7 +10,9 @@ conf:
...
+  kubernetes_nodes_base_box_id_libvirt: "zvfvrv/centos7"
```
Comment (ferrarimarco) [inline defaults.yaml]: No need to use your custom box, less things to maintain :)

**Comment by zvfvrv** [comment]
so do you think it's appropriate to insert a script to enable ssh password authentication?

**Comment by ferrarimarco** [comment]
You can do that inline, no need to add a script. See https://www.vagrantup.com/docs/provisioning/shell.html#inline

File: Vagrantfile [hunk 7]
```diff
@@ -347,6 +347,13 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
         end
         
         config.ssh.insert_key = false
+
+        $enablePswSSH = <<-SCRIPT
+        sudo sed -i "/^[^#]*PasswordAuthentication[[:space:]]no/c\PasswordAuthentication yes" /etc/ssh/sshd_config
```
Comment (ferrarimarco) [inline Vagrantfile]: What if `PasswordAuthentication `is not set at all? [The default is no](https://www.freebsd.org/cgi/man.cgi?sshd_config(5)).

File: Vagrantfile [hunk 8]
```diff
@@ -347,6 +347,13 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
         end
         
         config.ssh.insert_key = false
+
+        $enablePswSSH = <<-SCRIPT
```
Comment (ferrarimarco) [inline Vagrantfile]: Can you use a less cryptic name for this variable? Such as: `$enableSshPasswordAuthentication`

File: Vagrantfile [hunk 9]
```diff
@@ -347,6 +347,13 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
+        sudo sed -i "/^[^#]*PasswordAuthentication[[:space:]]no/c\PasswordAuthentication yes" /etc/ssh/sshd_config
+        sudo service sshd restart
+        SCRIPT
+
+        host.vm.provision "shell", inline: $enablePswSSH
```
Comment (ferrarimarco) [inline Vagrantfile]: This should run only for the box you use for libvirt. For the other one is not currently needed.

**Comment by ferrarimarco** [comment]
@zvfvrv can you please rebase against the latest master? Thanks

File: README.md [hunk 10]
```diff
@@ -52,7 +58,17 @@ you want to override any default setting, create `env.yaml` and save it in the
...
+
+In order to use libvirt as provider you need to set the variable `vagrant_provider` inside [`defaults.yaml`](defaults.yaml)
+
+      vagrant_provider: libvirt
+
+Vagrant needs to know that you want to use libvirt and not default VirtualBox. Then you can use the option `--provider=libvirt`, or you can setup environment variable
+
+      export VAGRANT_DEFAULT_PROVIDER=libvirt
+
+### Cleaning up
```
Comment (StefanoSalsano) [inline README.md]: chamge to Cleaning up and re-provisioning

File: Vagrantfile [hunk 11]
```diff
@@ -45,8 +45,10 @@ broadcast_address = n | (~n.instance_variable_get(:@mask_addr) & IPAddr::IN4MASK
 cluster_ip_cidr = settings["pod_network"]["cluster_ip_cidr"]
 service_ip_cidr = settings["pod_network"]["service_ip_cidr"]
 
+VAGRANT_PROVIDER = settings["conf"]["vagrant_provider"]
```
Comment (StefanoSalsano) [inline Vagrantfile]: we are now using lower case for the variables, so it should be "vagrant_provider"

**Comment by StefanoSalsano** [comment]
I've added two comments inside Readme.md and Vagrantfile

**Comment by ferrarimarco** [comment]
Apparently, you can run libvirt on Travis! https://github.com/jonashackt/vagrant-travisci-libvrt

This is going to be useful for CI. We can run the full suite end to end! (the virtualbox stuff will not be automatically tested tho)

File: .gitignore [hunk 12]
```diff
@@ -10,3 +10,5 @@ env.yaml
 kubernetes-playground-base.box
 
 ansible_output.txt
+
+.vscode
```
Comment (ferrarimarco) [inline .gitignore]: Missing newline at the end of the file

File: README.md [hunk 13]
```diff
@@ -52,6 +58,19 @@ you want to override any default setting, create `env.yaml` and save it in the
 same directory as the `defaults.yaml`. The [`Vagrantfile`](Vagrantfile) will
 instruct Vagrant to load it.
 
+#### Use Libvirt as provider
+
+In order to use libvirt as provider you need to set
```
Comment (ferrarimarco) [inline README.md]: You need to set the value of the `xxx` variable to `yyyyy`

File: README.md [hunk 14]
```diff
@@ -52,6 +58,19 @@ you want to override any default setting, create `env.yaml` and save it in the
...
+the variable `vagrant_provider` inside [`defaults.yaml`](defaults.yaml)
+
+      vagrant_provider: libvirt
+
+Vagrant needs to know that you want to use libvirt and not default VirtualBox.
+Then you can use the option `--provider=libvirt`,
+or you can setup environment variable
+
+      export VAGRANT_DEFAULT_PROVIDER=libvirt
```
Comment (ferrarimarco) [inline README.md]: No need to export this variable. Also, use `code tags` :)

File: Vagrantfile [hunk 15]
```diff
@@ -319,23 +321,41 @@ Vagrant.configure("2") do |config|
...
+          vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
+          vb.customize ["modifyvm", :id, "--natdnshostresolver2", "on"]
+          vb.gui = info[:show_gui]
+          vb.name = hostname
+        end
+      elsif(vagrant_provider == 'libvirt')
+        host.vm.provider :libvirt do |libvirt|
+          libvirt.cpus = info[:cpus]
+          libvirt.memory = info[:mem]
+          libvirt.nested = true
```
Comment (ferrarimarco) [inline Vagrantfile]: Can you set VM names as we do with the virtualbox provider?

File: Vagrantfile [hunk 16]
```diff
@@ -319,23 +321,41 @@ Vagrant.configure("2") do |config|
...
+          vb.customize ["modifyvm", :id, "--hwvirtex", "on"]
+          vb.customize ["modifyvm", :id, "--memory", info[:mem]]
+          vb.customize ["modifyvm", :id, "--name", hostname]
+          vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
+          vb.customize ["modifyvm", :id, "--natdnshostresolver2", "on"]
+          vb.gui = info[:show_gui]
+          vb.name = hostname
+        end
+      elsif(vagrant_provider == 'libvirt')
+        host.vm.provider :libvirt do |libvirt|
```
Comment (ferrarimarco) [inline Vagrantfile]: In virtualbox, we ask the VMs to use the host DNS resolver (by enabling `natdnshostresolver1` and `natdnshostresolver2`).

What about libvirt? Which DNS resolver are we using in this case?

File: Vagrantfile [hunk 17]
```diff
@@ -319,23 +321,41 @@ Vagrant.configure("2") do |config|
...
         host.vm.provision "shell" do |s|
           s.path = "scripts/linux/install-docker.sh"
           s.args = ["--user", "vagrant"]
         end
+        
+        config.ssh.insert_key = false
+        
+        if(vagrant_provider == 'libvirt')
+          $enableSshPasswordAuthentication = <<-'SCRIPT'
+          sudo grep -q "^PasswordAuthentication" /etc/ssh/sshd_config && sudo sed -i "s/^PasswordAuthentication no/PasswordAuthentication yes/" /etc/ssh/sshd_config || sudo sed -i -e "\$aPasswordAuthentication yes" /etc/ssh/sshd_config
```
Comment (ferrarimarco) [inline Vagrantfile]: Can you split this command in multiple lines?

Comment (ferrarimarco) [inline Vagrantfile]: Is `sudo` needed here? The shell provisioner runs scripts as privileged users [by default](https://www.vagrantup.com/docs/provisioning/shell.html#privileged)

File: defaults.yaml [hunk 18]
```diff
@@ -12,7 +12,10 @@ conf:
   minion_1_name: k8s-minion-1
   minion_2_name: k8s-minion-2
   minion_3_name: k8s-minion-3
-  kubernetes_nodes_base_box_id: "bento/centos-7.4"
+  vagrant_provider: libvirt
```
Comment (ferrarimarco) [inline defaults.yaml]: This should not be the default. Revert to virtualbox and eventually configure libvirt in your env.yaml.

File: Vagrantfile (hunk 15 above)
Comment (zvfvrv) [reply Vagrantfile]: yes i will set the default prefix to empty string

File: Vagrantfile (hunk 16 above)
Comment (zvfvrv) [reply Vagrantfile]: libvirt runs dnsmasq by default for DHCP and DNS services inside NAT virtual networks,
a dnsmasq instance for each virtual network

**Comment by ferrarimarco** [comment]
@zvfvrv great, we're there!

As our last step, can you attach a full log of the whole provisioning and configuration of the base box, the full environment, and which pods are running in all namespaces after the setup is done?

Thanks!

**Comment by zvfvrv** [comment]
@ferrarimarco please find attached the log of base-box building and the ansible output of the full environment.

[vagrant_base-box-builder.log](https://github.com/ferrarimarco/kubernetes-playground/files/4342777/vagrant_base-box-builder.log)
[ansible_output.txt](https://github.com/ferrarimarco/kubernetes-playground/files/4342778/ansible_output.txt)
[vagrant_prov_conf_env.log](https://github.com/ferrarimarco/kubernetes-playground/files/4342779/vagrant_prov_conf_env.log)


The pods running in all namespaces:
francesco@igor:~/Workspace/kubernetes-playground$ vagrant ssh k8s-master-1.k8s-play.local
[vagrant@k8s-master-1 ~]$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                                  READY   STATUS    RESTARTS   AGE
kube-system   calico-kube-controllers-7489ff5b7c-qfzbc              1/1     Running   0          4m51s
kube-system   calico-node-d9s2d                                     1/1     Running   0          4m44s
kube-system   calico-node-f7dgr                                     1/1     Running   0          4m44s
kube-system   calico-node-sh9zg                                     1/1     Running   0          4m44s
kube-system   calico-node-sp869                                     1/1     Running   0          4m51s
kube-system   coredns-6955765f44-5x825                              1/1     Running   0          4m51s
kube-system   coredns-6955765f44-8kmsz                              1/1     Running   0          4m51s
kube-system   etcd-k8s-master-1.k8s-play.local                      1/1     Running   0          5m4s
kube-system   kube-apiserver-k8s-master-1.k8s-play.local            1/1     Running   0          5m4s
kube-system   kube-controller-manager-k8s-master-1.k8s-play.local   1/1     Running   0          5m4s
kube-system   kube-proxy-69n8b                                      1/1     Running   0          4m44s
kube-system   kube-proxy-dhwxk                                      1/1     Running   0          4m44s
kube-system   kube-proxy-k7tjf                                      1/1     Running   0          4m51s
kube-system   kube-proxy-t9vvj                                      1/1     Running   0          4m44s
kube-system   kube-scheduler-k8s-master-1.k8s-play.local            1/1     Running   0          5m4s
[vagrant@k8s-master-1 ~]$

**Comment by ferrarimarco** [comment]
Great! Will merge ASAP

**Pull request merged**